### PR TITLE
Contract integration - Return values from inputs

### PR DESCRIPTION
**Pull request by jamesondh**
Approve and swap are still in progress.

### CODE REVIEW

File: webapp/src/assets/contractMethods.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+	'price_token_to_near_in',   
+	'price_token_to_near_out',  
+
+	'price_token_to_token_in',  
+	'price_token_to_token_out',  
+	
+	// Pool Info
+	'pool_info',
+	'get_pool',
```
Comment (robert-zaremba) [inline webapp/src/assets/contractMethods.js:15]: This function is private. Let's remove it from here.

File: webapp/src/assets/contractMethods.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	//Liquidity
+	'multi_balance_of'
+
+	];
+
+// TODO
+// Gas estimation
+// Approval transaction methods
+
+export let change_methods_lib_contract = [
```
Comment (robert-zaremba) [inline webapp/src/assets/contractMethods.js:28]: officially this is called `call` methods. So let's rename to `call_methods....`

File: webapp/src/assets/contractMethods.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	// pool
+	'create_pool',
+	'set_pool',
```
Comment (robert-zaremba) [inline webapp/src/assets/contractMethods.js:31]: `set_pool` is private, let's remove from here.

File: webapp/src/components/CurrencyTable.jsx [hunk 4]
```diff
@@ -29,33 +31,72 @@ export const CurrencyTable = () => {
...
+            await delay(1000).then(async function() {
+              try {
+                let allowance = await getAllowance(inputs.state.swap.from);
+                dispatch({ type: 'UPDATE_FROM_ALLOWANCE', payload: { allowance: allowance } });
+              } catch (e) {
+                console.error(e);
+              }
+            });
+          }
+        })();
```
Comment (robert-zaremba) [inline webapp/src/components/CurrencyTable.jsx:64]: let's move this function out to a modules scope. The function can take all necessary arguments as a parameter.

In general we should avoid making closures whenever possible. It's easy to create mess and lengthy not readable functions.

File: webapp/src/components/CurrencyTable.jsx [hunk 5]
```diff
@@ -29,33 +31,72 @@ export const CurrencyTable = () => {
...
       case 'input2':
         dispatch({ type: 'UPDATE_INPUT2_SELECTED_CURRENCY',
-          payload: { tokenIndex: newTokenIndex, logoUrl: newImageUrl, symbol: newSymbol, type: newType }
+          payload: { 
+            tokenIndex: newTokenIndex, 
+            logoUrl: newImageUrl, 
+            symbol: newSymbol, 
+            type: newType, 
+            address: newAddress }
         });
```
Comment (robert-zaremba) [inline webapp/src/components/CurrencyTable.jsx:95]: the `payload` is the same everywhere. Let's create it once before the switch, and reuse it.

File: webapp/src/components/Notification.jsx [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+import { NotificationContext } from "../contexts/NotificationContext";
+
+import Alert from 'react-bootstrap/Alert';
+
+export default function Notification(props) {
+
+  // Notification state
+  const notification = useContext(NotificationContext);
+  const { dispatch } = notification;
```
Comment (robert-zaremba) [inline webapp/src/components/Notification.jsx:11]: this line is not needed. Below we can write directly: `notification.dispatch`

File: webapp/src/contexts/InputsContext.js [hunk 7]
```diff
@@ -1,36 +1,47 @@
...
     from: {
       amount: "",     // Amount of tokens
       symbol: "",     // Symbol of token
       type: "",       // Native token (NEAR or ETH), ERC-20, NEP-21, ...
       logoUrl: "",    // Address of token logo image
-      tokenIndex: 1   // Index of token within token list
+      tokenIndex: 1,  // Index of token within token list
+      address: "",
+      isValid: false,
+      allowance: "",  // Required for NEP-21 -> ____ swap
```
Comment (robert-zaremba) [inline webapp/src/contexts/InputsContext.js:13]: Can we please create classes for some common objects? Here, all the content of `from`, `to`, `pool`, .... should be instance of a class to assure we have all fields.

File: webapp/src/contexts/InputsContext.js [hunk 8]
```diff
@@ -39,80 +50,211 @@ const initialState = {
...
+            type: state.swap.from.type,                 // leave type
+            logoUrl: state.swap.from.logoUrl,           // leave logo
+            tokenIndex: state.swap.from.tokenIndex,     // leave token index
+            address: state.swap.from.address,           // leave address
+            isValid: action.payload.isValid,            // UPDATE isValid
+            allowance: state.swap.from.allowance        // leave allowance
+          },                                            //
+          to: state.swap.to,                            // leave to input
+          needsApproval: state.swap.needsApproval,      // leave needsApproval
+          status: action.payload.status,                // UPDATE status
```
Comment (robert-zaremba) [inline webapp/src/contexts/InputsContext.js:80]: ditto, let's use class.

File: webapp/src/contexts/InputsContext.js [hunk 9]
```diff
@@ -39,80 +50,211 @@ const initialState = {
...
+          from: { 
+            amount: "",                                 // CLEAR amount
+            symbol: state.swap.from.symbol,             // leave symbol
+            type: state.swap.from.type,                 // leave type
+            logoUrl: state.swap.from.logoUrl,           // leave logo
+            tokenIndex: state.swap.from.tokenIndex,     // leave token index
+            address: state.swap.from.address,           // leave address
+            isValid: false,                             // RESET isValid
+            allowance: state.swap.from.allowance        // leave allowance
+          },                                            //
```
Comment (robert-zaremba) [inline webapp/src/contexts/InputsContext.js:208]: ditto

File: webapp/src/services/near-nep21-util.js [hunk 10]
```diff
@@ -0,0 +1,325 @@
+import { Contract} from 'near-api-js'
+import getConfig from '../config'
+
+const nearConfig = getConfig(process.env.NODE_ENV || 'development')
+const attachNear1 = '300000000000000';
+const attachNear2 = '60000000000000000000000';
```
Comment (robert-zaremba) [inline webapp/src/services/near-nep21-util.js:6]: Let's rename this variables. The first one I guess is `maxGas` the second one is `attachedNear`. Please also make sure that you are using them correctly (in the right order) in functions.

File: webapp/src/services/near-nep21-util.js [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+    owner_id: accountId, 
+    escrow_account_id: nearConfig.contractName });
+  console.log('Allowance: ', allowance);
+  return allowance;  
+}
+
+export async function gasCheck() {
+  // Set default
+  const near_limit = 0.6;
+  const bal = (await window.walletConnection.account().getAccountBalance()).available / 1000000000000000000000000;
```
Comment (robert-zaremba) [inline webapp/src/services/near-nep21-util.js:75]: the long number should be a constant. Please declare it a as `const NDENOM = 1e24` (we are using the same const in the backend -> near denomination). BTW - you can also use big num lib (`BN`) if precesion is important.

File: webapp/src/services/near-nep21-util.js [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+      if(ok)
+        val--;
+      res += x;
+    } else {
+      console.error("Error: Wrong Input");
+      return -1;
+    }
+  }
+  res = res + '0'.repeat(val);
+  return res;
```
Comment (robert-zaremba) [inline webapp/src/services/near-nep21-util.js:122]: this is far from optimal.
```
let start = 0;
for (; start< str.length; ++start) {
  if (str[start] !== '0')  break;
}
let end = str.length - 1;
for (; end> start; ++end) {
  if (str[from] !== '0')  break;
}
return str.slice(start,end+1)
```
Please also rename to `trimZeros`.

Comment (robert-zaremba) [reply webapp/src/services/near-nep21-util.js:122]: Btw, why do we remove zeros from end?

**Review by robert-zaremba — CHANGES_REQUESTED** [review]

File: webapp/src/services/near-nep21-util.js (hunk 12 above)
Comment (amityadav0) [reply webapp/src/services/near-nep21-util.js:122]: i thought there maybe any inconsistency in data, cleaning will result in less errors and to be on safe side as i got many unexpected errors when integrating UI

Comment (robert-zaremba) [reply webapp/src/services/near-nep21-util.js:113]: OK, but I still don't understand why do you want remove zeros from the end. `0012300` will become `123`. Shouldn't this be `12300`?

File: webapp/src/services/near-nep21-util.js [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+    if (str[start] !== '0')  break;
+  }
+  let end = str.length - 1;
+  for (; end> start; --end) {
+    if (str[end] !== '0')  break;
+  }
+  return str.slice(start,end+1)
+}
+
+function setAmount( value ) {
```
Comment (robert-zaremba) [inline webapp/src/services/near-nep21-util.js:94]: Please rename to `normalizeAmount` and add tests.

**Review by robert-zaremba — APPROVED** [review]
Approving, but before merging please solve the comment to rename the function and add few tests (at minimum covering a base example and ALL edge cases).
